src/start_server.ts

```diff
@@ -181,6 +161,179 @@ function startWithPort(userOptions: ServerOptions) {
+ * @param {string} contents Desired string contents of file
+ * @returns {Promise<T>}
+ */
+function writeFile(filename: string, contents: any) {
+  let p = new Promise((resolve, reject) => {
+    console.log('writing: ' + filename);
+    fs.writeFile(filename, contents, e => {
+      if (e) {
+        // just log error (don't reject)
+        console.warn('failed to write:' + filename, e);
+      }
+      resolve();
+    });
+  });
+  return p;
+}
+
+/**
+ * Reads a file asynchronously
+ * @param {string} filename Name of file to read
+ * @returns {Promise<string>} Promise of file contents
+ */
+function readFile(filename: string) {
+  let p = new Promise<string>((resolve, reject) => {
+    fs.readFile(filename, (err: any, data: any) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(data);
+      }
+    });
+  });
+  return p;
+}
+
+/**
+ * Gets the current TLS certificate (from current directory)
+ * or generates one if needed
+ * @returns {Promise<{}>} Promise of {serviceKey: string, certificate: string}
+ */
+function getTLSCertificate() {
+  let p = new Promise<{}>((resolve, reject) => {
+    let certificate: string;
+    let serviceKey: string;
+    readFile('.cert.pem')
+        .then((data: string) => { certificate = data; })
+        .then(() => readFile('.key.pem'))
+        .then((data: string) => { serviceKey = data; })
+        .then(() => resolve({
+          certificate: certificate,
+          serviceKey: serviceKey
+        }))
+        .catch(() => resolve(createTLSCertificate()));
+  });
+  return p;
+}
+
+/**
+ * Creates an HTTP(S) server
+ * @param app
+ * @param {ServerOptions} options
+ * @returns {Promise<http.Server>} Promise of server
+ */
+function createServer(app: any, options: ServerOptions) {
+  return new Promise<http.Server>((resolve, reject) => {
+
+    let protocol = options.protocol === 'http2' ? 'h2' : options.protocol;
+
+    if (isHttps(protocol)) {
+      getTLSCertificate()
+          .then((keys: any) => {
+            let opt = {
+              spdy: {protocols: [protocol]},
+              key: keys.serviceKey,
+              cert: keys.certificate
+            };
+            let server = http.createServer(opt, app);
+            resolve(server);
+          })
+          .catch(reject);
+    } else {
+      let spdyOptions = {protocols: [protocol], plain: true, ssl: false};
+      let server = http.createServer({spdy: spdyOptions}, app);
+      resolve(server);
+    }
+  });
+}
+
+/**
+ * Starts an HTTP(S) server on a specific port
+ * @param {ServerOptions} userOptions
+ * @returns {Promise<http.Server>} Promise of server
+ */
+function startWithPort(userOptions: ServerOptions) {
+  let options = applyDefaultOptions(userOptions);
+  let app = getApp(options);
+
+  let p = new Promise<http.Server>((resolve, reject) => {
+    createServer(app, options).then((server:any) => {
+      server.listen(options.port, options.hostname,
```

Comment:
Actually, I'd put the new Promise just around the server listening code, then you can remove the `catch()` below. Something like:

```
return createServer(app, options)
  .then((server) => new Promise((resolve, reject) => {
    server.listen(options.port, options.hostname, () => resolve(server));
    server.on('error', (e) => reject(e));
  })
  .then((_) => handleServerReady(options));
```